### PR TITLE
Fix SET command that sends DTX protocol command when shouldn't

### PR DESCRIPTION
**Pull request by huansong**
If there is an explicit BEGIN, we will begin transaction and setup DTX context on QEs at the time of the first SET command. So we need to add dtxSegments in SET to make sure we end the transaction at the time of END/COMMIT. This is required since b43629be3d84b00eb0e57dd285ba1da98ca1c27b where we no longer start DTX at the time of BEGIN.

However, we should not do this when there's NO explicit BEGIN. Because in that case, QEs will not have DTX context setup with the SET command, so if we decide to send DTX protocol command later (e.g. we have some direct-dispatch command that requires DTX with part of the QEs), we will send DTX protocol command to ALL dtxSegments. As a result, some QE might fail to recognize the DTX protocol command and throw error.

Fix the case by skipping adding dtxSegments when w/o explicit BEGIN/END. Also fix the comment in code and add test cases.

Dev pipeline: https://dev.ci.gpdb.pivotal.io/teams/main/pipelines/dtx-bug

## Here are some reminders before you submit the pull request
- [ ] Add tests for the change
- [ ] Document changes
- [ ] Communicate in the mailing list if needed
- [ ] Pass `make installcheck`
- [ ] Review a PR in return to support the community

### CODE REVIEW

File: src/test/regress/expected/distributed_transactions.out [hunk 1]
```diff
@@ -1044,3 +1044,36 @@ SELECT * FROM abc ORDER BY 1;
 RESET default_transaction_read_only;
  
 DROP TABLE abc;
```
Comment (soumyadeep2007) [inline src/test/regress/expected/distributed_transactions.out:1046]: Hmmm, what is weird is that I'm getting the same output if I run without your patch:

```
huansong=# create table tbl_dtx(a int, b int) distributed by (a);
insert into tbl_dtx values(1,1);

create or replace function dtx_set_bug()
  returns void
  language plpgsql
as $function$
begin
    execute 'update tbl_dtx set b = 1 where a = 1;';
    set optimizer=off;
end;
$function$;

set Test_print_direct_dispatch_info = true;
CREATE TABLE
INSERT 0 1
CREATE FUNCTION
SET
huansong=# begin;
set optimizer=false;
end;
BEGIN
SET
INFO:  Distributed transaction command 'Distributed Commit (one-phase)' to ALL contents: 0 1 2
COMMIT
huansong=# set optimizer=true;
SET
huansong=# select dtx_set_bug();
INFO:  (slice 0) Dispatch command to ALL contents: 0 1 2
INFO:  Distributed transaction command 'Distributed Prepare' to ALL contents: 0 1 2
INFO:  Distributed transaction command 'Distributed Commit Prepared' to ALL contents: 0 1 2
 dtx_set_bug 
-------------
 
(1 row)
```

**I think we aren't direct dispatching here.**

However, the one in your commit message does fail and does show evidence of direct dispatch:
```
huansong=# create table t1(a int, b int);

CREATE OR REPLACE FUNCTION dtx_bug()
  RETURNS void
  LANGUAGE plpgsql
AS $function$
BEGIN
    EXECUTE 'UPDATE t1 set b = 1 where a = 1 or a = 2;';
    set optimizer=off;
END;
$function$;
SELECT dtx_bug();
NOTICE:  Table doesn't have 'DISTRIBUTED BY' clause -- Using column named 'a' as the Greenplum Database data distribution key for this table.
HINT:  The 'DISTRIBUTED BY' clause determines the distribution of data. Make sure column(s) chosen are the optimal data distribution key to minimize skew.
ERROR:  relation "t1" already exists
INFO:  Distributed transaction command 'Distributed Prepare' to ALL contents: 0 1 2
INFO:  Distributed transaction command 'Distributed Commit Prepared' to ALL contents: 0 1 2
CREATE FUNCTION
INFO:  (slice 0) Dispatch command to PARTIAL contents: 1 0
INFO:  Distributed transaction command 'Distributed Commit (one-phase)' to ALL contents: 1 0 2
INFO:  Distributed transaction command 'Distributed Abort (No Prepared)' to ALL contents: 1 0 2
ERROR:  Distributed transaction 352319 not found (cdbtm.c:2204)  (seg2 10.0.0.204:7004 pid=3891917) (cdbtm.c:2204)

```

We need to see this kind of output: `INFO:  (slice 0) Dispatch command to PARTIAL contents: 1 0`

So we will have to tweak the test case somehow to ensure direct dispatch.

**Comment by SmartKeyerror** [comment]
I'm wondering why PR #7529 add the SET command into two-phase commit transaction. 

Generally, the SET command will dispatch all QEs, including the reader gang and the writer gang, as Heikki said it doesn't change anything on disk at [issuecomment-450806538](https://github.com/greenplum-db/gpdb/pull/6596#issuecomment-450806538).

Maybe we could just remove it:

```
--- a/src/backend/cdb/dispatcher/cdbdisp_query.c
+++ b/src/backend/cdb/dispatcher/cdbdisp_query.c
@@ -329,7 +329,6 @@ CdbDispatchSetCommand(const char *strCommand, bool cancelOnError)
 
                cdbdisp_dispatchToGang(ds, rg, -1);
        }
-       addToGxactDtxSegments(primaryGang);
 
        /*
         * No need for two-phase commit, so no need to call
```

What's your thought?

**Comment by huansong** [comment]
> I'm wondering why PR https://github.com/greenplum-db/gpdb/pull/7529 add the SET command into two-phase commit transaction.

@SmartKeyerror I wondered the same thing during investigation. It was actually pointed out during the review of that PR: https://github.com/greenplum-db/gpdb/pull/7529/files#r285830213 . And I tested that w/o `addToGxactDtxSegments` we would have issue with cases like:
```sql
create temp table tt(a int);

-- trx1
begin;
set optimizer=default;
commit;

-- trx2
begin;
select * from tt;
end;
-- trx2 will throw an error:
-- ERROR:  Unexpected internal error (assert.c:48)  (seg0 127.0.1.1:7002 pid=30472) (assert.c:48)
-- DETAIL:  FailedAssertion("!(gxid == getDistributedTransactionId())", File: "cdbtm.c", Line: 2056)
```
This is because the first `SET` command in `trx1` is going to set up DTX context on every QE (see https://github.com/greenplum-db/gpdb/blob/76e3f182da0bd041e12d81c78861901a4b39d1f6/src/backend/cdb/cdbtm.c#L1841-L1851) , which will start the transaction. Once this is done, the QE will expect a DTX protocol command to be sent to itself in order to end the transaction and clear its DTX related variables. Otherwise, the QE is going to have some weird limbo state which would cause various problem for subsequent transactions.

Basically, `addToGxactDtxSegments` doesn't mean 2PC is used, it only means "DTX is used" and QD is going to send a DTX protocol command to the QEs involved in the transaction. And for explicit `BEGIN`, DTX is always used. Then, some other logic will be used to decide whether 1PC or 2PC is used. If we only have `SET` command in the transaction, only 1PC  is used.

**Review by SmartKeyerror — APPROVED** [review]
